src/Processors/Streaming/ReplayStreamTransform.cpp

```diff
@@ -16,34 +16,145 @@ namespace Streaming
 
 constexpr Int64 MAX_WAIT_INTERVAL_MS = 500;
 
-ReplayStreamTransform::ReplayStreamTransform(const Block & header, Float32 replay_speed_, Int64 last_sn_)
+ReplayStreamTransform::ReplayStreamTransform(const Block & header, Float32 replay_speed_, Int64 last_sn_, const String & replay_time_col_)
     : ISimpleTransform(header, header, true, ProcessorID::ReplayStreamTransformID)
     , replay_speed(replay_speed_)
     , last_sn(last_sn_)
     , enable_replay(last_sn >= 0)
+    , replay_time_col(replay_time_col_)
 {
-    append_time_index = header.getPositionByName(ProtonConsts::RESERVED_APPEND_TIME);
+    time_index = header.getPositionByName(replay_time_col);
+    auto & type = header.getByPosition(time_index).type;
+    /// user defined replay_time_col must be DateTime64
+    if (replay_time_col != ProtonConsts::RESERVED_APPEND_TIME && !isDateTime64(type))
+        throw Exception(
+            fmt::format("ReplayStreamTransform need datatime64 type column, but got {}", type->getName()), ErrorCodes::LOGICAL_ERROR);
     sn_index = header.getPositionByName(ProtonConsts::RESERVED_EVENT_SEQUENCE_ID);
 }
 
-void ReplayStreamTransform::transform(Chunk & chunk)
+/**
+ * @brief put the row that has different time into different chunk.
+ * For example: input_chunk: [1, 1, 1, 2, 2, 2, 3, 3, 3]
+ * After the cutChunk, the output_chunks will be [[1, 1, 1], [2, 2, 2], [3, 3, 3]]
+ */
+void ReplayStreamTransform::cutChunk(Chunk & input_chunk)
 {
-    if (!enable_replay || !chunk.rows())
+    assert(input_chunk.rows() > 0);
+    size_t index = 0;
+    size_t cur_index = 0;
+    auto & columns = input_chunk.getColumns();
+
+    auto cut_into_chunks = [&](size_t start_pos, size_t end_pos) {
+        Chunk chunk;
+        for (const auto & col : columns)
+            chunk.addColumn(col->cut(start_pos, end_pos - start_pos));
+        output_chunks.push(Chunk());
+        output_chunks.back().swap(chunk);
+    };
+    while (index < input_chunk.rows())
+    {
+        if (columns[time_index]->getInt(index) != columns[time_index]->getInt(cur_index))
+        {
+            cut_into_chunks(cur_index, index);
+            cur_index = index;
+        }
+        ++index;
+    }
+
+    cut_into_chunks(cur_index, index);
+    input_chunk.clear();
+}
+
+void ReplayStreamTransform::work()
+{
+    if (input_data.exception)
+    {
+        /// Skip transform in case of exception.
+        output_data = std::move(input_data);
+        has_input = false;
+        has_output = true;
+        return;
+    }
+
+    /// proton: starts.
+    auto start_ns = MonotonicNanoseconds::now();
+    metrics.processed_bytes += input_data.chunk.bytes();
+    /// proton: ends.
+    try
+    {
+        if (enable_replay && input_data.chunk.rows() && output_chunks.empty())
+            cutChunk(input_data.chunk);
+        transform(input_data.chunk, output_data.chunk);
```

Comment:
after cutChunk , the input_data.chunk was cleared, why not move the `cutChunk` into `transform()` ?